packages/skygear-core/src/imageprocessing.ts

```diff
@@ -0,0 +1,162 @@
+import { encodeQueryComponent } from "./url";
+
+/**
+ * @public
+ */
+export interface ImageProcessingPipelineBuilderResizeOptions {
+  scalingMode?: "lfit" | "mfit" | "fill" | "pad" | "fixed";
+  targetWidth?: number;
+  targetHeight?: number;
+  longerSide?: number;
+  shorterSide?: number;
+  color?: string;
+}
+
+type Operation = FormatOperation | QualityOperation | ResizeOperation;
+
+interface FormatOperation {
+  op: "format";
+  format: "jpg" | "png" | "webp";
+}
+
+interface QualityOperation {
+  op: "quality";
+  absoluteQuality: number;
+}
+
+interface ResizeOperation extends ImageProcessingPipelineBuilderResizeOptions {
+  op: "resize";
+}
+
+/**
+ * @public
+ */
+export class ImageProcessingPipelineBuilder {
+  private _ops: Operation[];
+
+  constructor() {
+    this._ops = [];
+  }
+
+  format(format: "jpg" | "png" | "webp"): this {
+    this._ops.push({
+      op: "format",
+      format,
+    });
+    return this;
+  }
+
+  quality(absoluteQuality: number): this {
+    this._ops.push({
+      op: "quality",
+      absoluteQuality,
+    });
+    return this;
+  }
+
+  resize(options: ImageProcessingPipelineBuilderResizeOptions): this {
+    this._ops.push({
+      op: "resize",
+      ...options,
+    });
+    return this;
+  }
+
+  getName(): string {
+    return "pipeline";
+  }
+
+  getValue(): string {
```

Comment:
I think renaming `ImageProcessingPipelineBuilder` to `ImageProcessingPipeline` and `getValue` to `toString` would be better. This class is actually the pipeline itself, and the 'value' is the string representation of the pipeline.

Reply:
I think the Builder suffix better communicate the purpose.

For `toString`, I have thought about it when I wrote the spec. If it is named `toString`, should the return value be `pipeline=image/...` or `image/...`? Naming it toString would bring such ambiguity so I finally settled with calling it `getValue`